### PR TITLE
🏎 Lazily install "bower" if required

### PR DESCRIPTION
**Pull request by Turbo87**
/cc @stefanpenner @rwjblue @nathanhammond

### CODE REVIEW

File: lib/tasks/bower-install.js [hunk 1]
```diff
@@ -3,68 +3,97 @@
...
+              message.data.name + ' ' + pick.endpoint.target);
+          });
+        } else if (message.level === 'info' && options.verbose) {
+          // e.g.
+          //   cached git://example.com/some-package.git#1.0.0
+          ui.writeLine('  ' + chalk.green(message.id) + ' ' + message.message);
+        }
       }
-    }
+    }.bind(this));
```
Comment (rwjblue) [inline lib/tasks/bower-install.js]: Perhaps `() => {}`?

File: package.json [hunk 2]
```diff
@@ -37,7 +37,6 @@
   },
   "dependencies": {
     "amd-name-resolver": "0.0.6",
-    "bower": "^1.3.12",
     "bower-config": "^1.3.0",
```
Comment (rwjblue) [inline package.json]: Should do the same treatment for the other bower packages (bower-config, bower-endpoint-parser)?

File: lib/tasks/bower-install.js [hunk 3]
```diff
@@ -3,68 +3,97 @@
...
+    });
+  },
+
+  installBower: function() {
+    var ui = this.ui;
+    var chalk = require('chalk');
+
+    ui.startProgress(chalk.green('NPM: Installing bower ...'));
+
+    return Promise.cast(execa('npm', ['install', 'bower@^1.3.12'], {
```
Comment (rwjblue) [inline lib/tasks/bower-install.js]: > Promise.cast is dead, don't use it

- @stefanpenner in https://github.com/emberjs/ember.js/issues/13984

Comment (Turbo87) [inline lib/tasks/bower-install.js]: https://github.com/ember-cli/ember-cli/blob/9b3387123e84d0c05775490f1903e7c6f11f7b38/lib/utilities/npm.js#L21 ... ups 😉

File: lib/tasks/bower-install.js (hunk 1 above)
Comment (Turbo87) [inline lib/tasks/bower-install.js]: yeah, I'll use that in the future. wanted to keep this compatible with the other stuff for now.

File: package.json (hunk 2 above)
Comment (Turbo87) [inline package.json]: I thought about that, and decided to tackle the biggest issue/dependency (`bower`) first. In the future we can and should make those optional too.

**Comment by Turbo87** [comment]
> I'm also curious how this interacts with blueprints using addBowerPackageToProject?

I've tested this by running `ember install ember-cli-mocha@0.12.0` in a fresh project with a linked CLI and it works quite well from what I can tell so far.

> I'm not sure doing nothing (if bower.json doesn't exist) is right?

That is the same logic that we've used before. I didn't change that part of the code, but I agree that we should probably have a look at that part too. Seems unrelated to this change though.

**Comment by rwjblue** [comment]
> That is the same logic that we've used before. I didn't change that part of the code, but I agree that we should probably have a look at that part too. Seems unrelated to this change though.

Can you make an issue for that (so we don't loose track)? I suspect if we can get this right then we can remove bower.json from the default app/addon blueprint...

File: lib/tasks/bower-install.js (hunk 3 above)
Comment (rwjblue) [inline lib/tasks/bower-install.js]: Haha, well we can fix this one usage here (since we have to change it anyways) and then go on a hunt...

Comment (Turbo87) [inline lib/tasks/bower-install.js]: changed `Promise.cast()` to `Promise.resolve()` and apparently it is still working well 😀

File: lib/tasks/bower-install.js [hunk 4]
```diff
@@ -3,68 +3,97 @@
 // Runs `bower install` in cwd
 var path       = require('path');
 var existsSync = require('exists-sync');
+var execa = require('execa');
 var Promise    = require('../ext/promise');
 var Task       = require('../models/task');
 var formatPackageList = require('../utilities/format-package-list');
+var resolve = Promise.denodeify(require('resolve'));
```
Comment (stefanpenner) [inline lib/tasks/bower-install.js]: thoughts on `resolve.sync` ?

File: package.json (hunk 2 above)
Comment (stefanpenner) [inline package.json]: 👍

File: lib/tasks/bower-install.js (hunk 4 above)
Comment (Turbo87) [inline lib/tasks/bower-install.js]: why use sync when we can use the async version without blocking the progress indicator?

Comment (rwjblue) [inline lib/tasks/bower-install.js]: Ya, thats a good enough reason to use async version to me. 😺

**Comment by rwjblue** [comment]
@homu r+

File: package.json (hunk 2 above)
Comment (Glarregle) [reply package.json:40]: Coming from the future.. Was wondering why `ember-cli` still has `bower-config` has dependency.

**Pull request merged**